packages/conform-zod/v4/coercion.ts

```diff
@@ -0,0 +1,415 @@
+	if (typeof value !== 'string') {
+		return value;
+	}
+
+	return value.trim() === '' ? value : Number(value);
+}
+
+function coerceBoolean(value: unknown) {
+	if (typeof value !== 'string') {
+		return value;
+	}
+
+	return value === 'on' ? true : value;
+}
+
+function coerceDate(value: unknown) {
+	if (typeof value !== 'string') {
+		return value;
+	}
+
+	const date = new Date(value);
+
+	// z.date() does not expose a quick way to set invalid_date error
+	// This gets around it by returning the original string if it's invalid
+	// See https://github.com/colinhacks/zod/issues/1526
+	if (isNaN(date.getTime())) {
+		return value;
+	}
+
+	return date;
+}
+
+function coerceBigInt(value: unknown) {
+	if (typeof value !== 'string') {
+		return value;
+	}
+
+	if (value.trim() === '') {
+		return value;
+	}
+
+	try {
+		return BigInt(value);
+	} catch {
+		return value;
+	}
+}
+
+/**
+ * @deprecated Conform strip empty string to undefined by default
+ */
+export function ifNonEmptyString(fn: (text: string) => unknown) {
+	return (value: unknown) => {
+		let result = stripEmptyString(value);
+
+		if (typeof result === 'string') {
+			result = fn(result);
+		}
+
+		return result;
+	};
+}
+
+export type DefaultCoercionType =
+	| 'string'
+	| 'file'
+	| 'number'
+	| 'boolean'
+	| 'date'
+	| 'bigint';
+
+export type CoercionFunction = (value: unknown) => unknown;
+
+function compose(
+	a: CoercionFunction,
+	b: CoercionFunction,
+	c: CoercionFunction = (i) => i,
+): CoercionFunction {
+	return (value) => c(b(a(value)));
+}
+
+function selectDefaultCoercion(
+	type: $ZodType,
+	defaultCoercion: Record<DefaultCoercionType, CoercionFunction>,
+): CoercionFunction | null {
+	const def = type._zod.def;
+
+	if (
+		def.type === 'string' ||
+		def.type === 'literal' ||
+		def.type === 'enum' // || def.type === 'nativeEnum'
+	) {
+		return defaultCoercion.string;
+	} else if (def.type === 'file') {
+		return defaultCoercion.file;
+	} else if (def.type === 'number') {
+		return defaultCoercion.number;
+	} else if (def.type === 'boolean') {
+		return defaultCoercion.boolean;
+	} else if (def.type === 'date') {
+		return defaultCoercion.date;
+	} else if (def.type === 'bigint') {
+		return defaultCoercion.bigint;
+	}
+
+	return null;
+}
+
+/**
+ * Reconstruct the provided schema with additional preprocessing steps
+ * This strips empty values to undefined and coerces string to the correct type
+ */
+export function enableTypeCoercion<Schema extends $ZodType>(
+	type: Schema,
+	options: {
+		cache: Map<$ZodType, $ZodType>;
+		coerce: (type: $ZodType) => CoercionFunction | null;
+		stripEmptyValue: CoercionFunction;
+	},
+): $ZodType {
+	const result = options.cache.get(type);
+
+	// Return the cached schema if it's already processed
+	// This is to prevent infinite recursion caused by z.lazy()
+	if (result) {
+		return result;
+	}
+
+	let schema: $ZodType = type;
+	const zod = (type as unknown as $ZodTypes)._zod;
+	const def = zod.def;
+	const constr = zod.constr;
```

Comment:
The original schema defined by the user inside Zod is stored in `constrt`. `transform` is used to redefine the original schema by applying schema operations to the original schema.